### ISSUE TITLE
Clock sync via Device Time Request

### ISSUE THREAD

**Issue by KestrelLuca** [MicrochipTech/atsamr34_lorawan_rn_parser#11]
Hello,
Thanks for providing this useful example.

I'm trying to implement in the app layer the LoRa mac command "Device Time Request".
From the lorawan.h in the `typedef enum _LorawanAttributes` I can see the attribute:

`    /* If set, ED shall send DeviceTimeReq cmd in next TX */`
   `  SEND_DEVICE_TIME_CMD,`

From the above comment, I've understood that I should trigger a "mac **set** DevTimeReq ".
Obviously the entry `{"DevTimeReq",  NULL, Parser_LoraGetDevTime, 0,0}` has been added to the `static const parserCmdEntry_t maParserLoraSetCmd[] ` array
and I wrote the following `Parser_LoraGetDevTime ()` function:

`void Parser_LoraGetDevTime(parserCmdInfo_t* pParserCmdInfo)` 
`{` 
`	StackRetStatus_t status;` 
`	status = LORAWAN_SetAttr(SEND_DEVICE_TIME_CMD, NULL);` 
`	pParserCmdInfo->pReplyCmd = (char*)gapParserLorawanStatus[status];	`  
`}` 

The command seems to be accepted by the stack since the returned status is  `LORAWAN_SUCCESS`.

The problem is that after triggering the "mac set DevTimeReq" I'm not able to send any uplink (can't see them on the TTN device page) therefore the end device neither is sending the DeviceTimeReq cmd.

What I'm missing here?
Is there any other document I should read other than the "SAM R34/R35 Microchip LoRaWAN™ Stack Software API
Reference Manual"?

Best regards
Luca

**Comment by KestrelLuca** [comment]
Update:
Migrating my devices, app and gateway to the TTN V3, I'm now able to see my DeviceTimeReq being correctly sent (on the next tx). The DeviceTimeAns is also there with the expected data.
The problem is that I don't know how to handle this frame simply because I can't see it being received from the device.

I've placed a BP in the `ParserAppData()` callback (into parser_lorawan.c) but this isn't the right place to break the execution (I guess). 
Is the received time stored somewhere else in the stack?
In the example, I can see the attribute `DEVICE_GPS_EPOCH_TIME` is this related to the DeviceTimeAns?

![image](https://user-images.githubusercontent.com/75615127/118508920-851fe680-b727-11eb-9b38-a59fb594cc45.png)

**Comment by gd91** [comment]
DEVICE_GPS_EPOCH_TIME is received in DeviceTimeAnswer MAC command
For further support please use [www.microchip.com/support](www.microchip.com/support)

**Comment by KestrelLuca** [comment]
Hi @gd91, thanks for your reply,
you are quoting the "DeviceTimeAnswer MAC command" from the LoRaWAN spec (first introduced in v1.0.3.)

The issue here is related to the MLS where the DEVICE_GPS_EPOCH_TIME is the name of the MAC stack attribute (in the project search for _LorawanAttributes in the lorawan.h file ).

The problem is that the DEVICE_GPS_EPOCH_TIME attribute isn't even listed in the latest MLS stack documentation $2.3.2 [https://ww1.microchip.com/downloads/en/DeviceDoc/SAM-R34-R35-Microchip-LoRaWAN-Stack-Software-API-Reference-Manual-DS70005382B.pdf](SAM R34/R35 Microchip LoRaWAN™ Stack Software API
Reference Manual) 

The "DeviceTimeRequest & DeviceTimeAnswer MAC command" are not implemented in this example but this is not a problem, I can do it but I need the information ready in the MLS stack so I can retrieve them using the  LORAWAN_GetAttr() API.
And this is what I did:
- prepare the DeviceTimeReq (done the the `Parser_LoraGetDevTime()` previously posted
- Send an uplink (as explained in the comment for the SEND_DEVICE_TIME_CMD attribute `  /* If set, ED shall send DeviceTimeReq cmd in next TX */`)
- Then I tried to get the DEVICE_GPS_EPOCH_TIME and the stack is returning wrong data.

I wish Microchip support was able to help. I've contacted them before opening an issue here and still I'm waiting for some help.
